src/command/archive/push/file.c

```diff
@@ -163,11 +163,39 @@ archivePushFile(
             const String *const walSegmentChecksum = strNewEncode(
                 encodingHex, pckReadBinP(ioFilterGroupResultP(ioReadFilterGroup(read), CRYPTO_HASH_FILTER_TYPE)));
 
+            // Is the check command currently running?
+            const bool checkLockHeld = lockRead(
+                cfgOptionStr(cfgOptLockPath), cfgOptionStr(cfgOptStanza), lockTypeCheck).status == lockReadStatusValid;
+
+            // Is the backup command running for a specific repo?
+            const Variant *runningBackupRepoIdx = lockRead(
+                cfgOptionStr(cfgOptLockPath), cfgOptionStr(cfgOptStanza), lockTypeBackup).data.repoIdx;
+
             // Check each repo for the WAL segment
             for (unsigned int repoListIdx = 0; repoListIdx < lstSize(repoList); repoListIdx++)
             {
                 const ArchivePushFileRepoData *const repoData = lstGet(repoList, repoListIdx);
 
+                // Skip archiving if archive retention is set to 0 and the backup command isn't currently running for this specific
+                // repository. Archiving should always happen when the check command is running for all the repositories.
+                bool lockHeld = false;
+                if (checkLockHeld || (runningBackupRepoIdx != NULL && varUInt(runningBackupRepoIdx) == repoData->repoIdx))
+                    lockHeld = true;
+
+                if (!lockHeld
+                    && (cfgOptionIdxTest(cfgOptRepoRetentionArchive, repoData->repoIdx)
+                        && cfgOptionIdxUInt(cfgOptRepoRetentionArchive, repoData->repoIdx) == 0))
```

Comment:
Is the cfgOptionIdxTest() required here? cfgOptRepoRetentionArchive should always be valid for archive-push.

Reply:
Indeed :+1:

Reply:
Hm, I did try to remove the Test() but then I get `ASSERT: [025]: option 'repo1-retention-archive' is null but non-null was requested`. So I at least need to test if not null before trying to check if = 0.
See [3b806c6](https://github.com/pgbackrest/pgbackrest/pull/1926/commits/3b806c6233f14888c44cde32de34c7455f24562e).

Reply:
OK, yeah I see. The default is usually set in cfgLoadUpdateOption() but archive-push does not have all the info needed to set the default.